src/t8_forest/t8_forest_ghost.cxx

```diff
@@ -1929,4 +1880,348 @@ t8_forest_ghost_destroy (t8_forest_ghost_t *pghost)
+  /* Store any user data that may reside on the forest */
+  store_user_data = t8_forest_get_user_data (forest);
+  /* Set the user data for the search routine */
+  t8_forest_set_user_data (forest, &data);
+  /* Loop over the trees of the forest */
+  t8_forest_search (forest, search_fn, NULL, NULL);
+
+  /* Reset the user data from before search */
+  t8_forest_set_user_data (forest, store_user_data);
+
+  /* Reset the data arrays */
+  sc_array_reset (&data.face_owners);
+  sc_array_reset (&data.bounds_per_level);
+#ifdef T8_ENABLE_DEBUG
+#endif
+}
+
+t8_forest_ghost_face::t8_forest_ghost_face (int version)
+  : t8_forest_ghost_w_search (T8_GHOST_FACES, t8_forest_ghost_search_boundary), version (version)
+{
+  T8_ASSERT (1 <= version && version <= 3);
+}
+
+void
+t8_forest_ghost_face::search_for_ghost_elements (t8_forest_t forest)
+{
+  t8_global_productionf (" t8_forest_ghost_face::step_2 \n");
+  T8_ASSERT (forest->ghosts != NULL);
+  t8_forest_ghost_t ghost = forest->ghosts;
+  if (version == 3) {
+    t8_global_productionf ("t8_forest_ghost_face::step_2: t8_forest_ghost_fill_remote_v3(forest)\n");
+    t8_forest_ghost_fill_remote_v3 (forest);
+  }
+  else {
+    /* Construct the remote elements and processes. */
+    t8_global_productionf (
+      "t8_forest_ghost_face::step_2: t8_forest_ghost_fill_remote (forest, ghost, ghost_version != 1)\n");
+    t8_forest_ghost_fill_remote (forest, ghost, version != 1);
+  }
+}
+
+/* Wrapper for derived face class */
+t8_forest_ghost_interface_c *
+t8_forest_ghost_interface_face_new (int version)
+{
+  t8_debugf ("Call t8_forest_ghost_interface_face_new.\n");
+  T8_ASSERT (1 <= version && version <= 3);
+  t8_forest_ghost_face *ghost_interface = new t8_forest_ghost_face (version);
+  return (t8_forest_ghost_interface_c *) ghost_interface;
+}
+
+t8_forest_ghost_interface_c *
+t8_forest_ghost_interface_stencil_new ()
+{
+  t8_debugf ("Call t8_forest_ghost_interface_stencil_new.\n");
+  t8_forest_ghost_stencil *ghost_interface = new t8_forest_ghost_stencil ();
+  return (t8_forest_ghost_interface_c *) ghost_interface;
+}
+
+int
+t8_forest_ghost_interface_face_version (t8_forest_ghost_interface_c *ghost_interface)
+{
+  T8_ASSERT (ghost_interface != NULL);
+  T8_ASSERT (ghost_interface->t8_ghost_get_type () == T8_GHOST_FACES);
+  t8_forest_ghost_face *ghost_interface_passed = (t8_forest_ghost_face *) ghost_interface;
+
+  return ghost_interface_passed->get_version ();
+}
+
+/**
+ * Derived class for a stencil on an uniform mesh.
+ */
+
+void
+t8_forest_ghost_stencil::do_ghost (t8_forest_t forest)
+{
+  /**
+   * Compute bounds for elements
+   */
+
+  t8_forest_ghost_init (&forest->ghosts, ghost_type);
+
+  t8_locidx_t current_index;                   // counter over all local elements
+  t8_locidx_t ielement, num_elements_in_tree;  // count over the local elements in a tree
+  t8_eclass_scheme_c *eclass_scheme;
+  t8_eclass_t tree_class;
+  const t8_element_t *element;
```

Comment:
I know that most of the other functions also declare variables like this, but this is a very old and bad coding style. This keeps one from using as much const as possible. It would be better to declare variables at the position where they are assigned.